Rule 2ee8b8 ("Visible label is part of accessible name"): introducing a new "label in name algorithm". (#2075)
auto-wcag/auto-wcag: pages/glossary/label-in-name-algorithm.md

---
title: Label in Name Algorithm
key: label-in-name-algorithm
unambiguous: true
objective: false
input_aspects:
  - CSS styling
  - DOM tree
---

> 

Reviewer: I think we should add a starting bit in the line of:
```
To check whether an [element][] has <dfn>its label contained in its name</dfn>, follow this algorithm:
```

Author: Done in https://github.com/act-rules/act-rules.github.io/pull/2075/commits/46294ddee604791751018a48462b42c5bd3900ec

- For each character that either a) represents non-text content, or b) isn't a letter or a digit: replace that character with a space character.
    - For a) Judgement of "non-text" probably can't be fully automated. eg. "X" for "close" probably can be, but presumably there are more cases than this.
    - For b) Use the unicode classes Letter, Mark, and "Number, Decimal Digit [Nd]". (This will exclude hyphens, punctuation, emoji, and more.)
- Remove all characters that are within parentheses (AKA round brackets).
    - Ignore square brackets and braces.
- Split the string into a list of strings, using a whitespace regular expression as the separator.
    - This 'split' operation must:
        - Effectively remove leading and trailing whitespace as a pre-processing step.
        - If the string was all whitespace before this operation: result in an empty list.

Then do the check: is the tokenized 'label' a sublist of the tokenized 'name'?
- This 'sublist' check has these properties:
    - It checks whether elements are consecutive or not. i.e. it checks for a substring, in the computer science sense of the term. Not a subsequence.
    - An empty list is a sublist of any list.

> If the answer is "yes" (i.e. the tokenized 'label' a sublist of the tokenized 'name'), then the target element passes the rule.  Otherwise, it fails the rule.

Reviewer: This should not refer to "the rule" since definitions are supposed to live on their own.

Author: I agree.   I fixed it in this commit a minute ago: https://github.com/act-rules/act-rules.github.io/pull/2075/commits/24d0ffc340dbcc44ad2bf7ca9cb6879610166227

---
title: Label in Name Algorithm
key: label-in-name-algorithm
unambiguous: true
objective: false
input_aspects:
  - CSS styling
  - DOM tree
---

To check whether an [element][] has <dfn>its label contained in its name</dfn>, follow this algorithm:

Let 'label' be the [visible inner text][] of the target element.  Let 'name' be the [accessible name][] of the target element.  Both 'label' and 'name' are strings.

Sub-algorithm to tokenize a string:

- Convert the string to lower case.

> - For each character that either a) represents non-text content, or b) isn't a letter or a digit: replace that character with a space character.

Reviewer 2: I think this can cause false positives with words that are not consistently hyphenated. For example:

'antimatter' vs 'anti-matter' (the [Wikipedia antimatter](https://en.wikipedia.org/wiki/Antimatter) article uses both)
'derisk' vs 'de-risk' (Cambridge dictionary uses first spelling, Collins dictionary uses the second)
'nonnegative' vs 'non-negative' https://math.stackexchange.com/a/3344027

Author: You're right, and I assume the voice control software would handle it fine, but if it was a customer who was arguing this, I would say "too bad" (diplomatically, of course).   This is part of a larger issue that I call "you need to play ball with the automated check".   I run into customer arguments like this often enough that I have a "canned response" it.  It reads, in part: "It's not feasible for an automated checker to know any better in this situation.  When you use an automated checker instead of a manual audit, in order to gain efficiency, you lose accuracy.  This is an unavoidable example of that." 

For the page author to "play ball" for "label in name" would be easy: use the same spelling, including hyphenation, in the aria-label that they used in the visible label.  This is, I think, not a lot to ask of a page author.

I could add an assumption to the rule.  Something along the lines of "This rule assumes that for any word (including any hyphenated word) that appears in both the accessible name and the visible label, the same spelling and hyphenation is used in both places.  For example, using 'antimatter' in the accessible name and 'anti-matter' in the visible label would fail this rule, but arguably pass the Success Criterion.

Reviewer 2: Bearing in mind this is for voice control, would a voice control user pronounce the hyphenated and non-hyphenated versions differently? Personally, I would pronounce both versions identically:

'antimatter' vs 'anti-matter' 
'derisk' vs 'de-risk' 
'nonnegative' vs 'non-negative'

and sometimes it's context specific:

'1-2' might be pronounced as 'one two' in  a sports context, but as 'one minus two' in a maths context/

Reviewer 3: > Bearing in mind this is for voice control, would a voice control user pronounce the hyphenated and non-hyphenated versions differently? Personally, I would pronounce both versions identically:
> 
> 'antimatter' vs 'anti-matter' 'derisk' vs 'de-risk' 'nonnegative' vs 'non-negative'

It's worth noting that Dragon voice recognition software only recognises and responds to words that are in its dictionary. For instance, if the accessible name is 'nonnegative', nothing you say will cause Dragon to interact with it if 'nonnegative' is not in its dictionary. If the accessible name is 'non negative', you could say either or both words because both will be in the dictionary. I suspect the same would be the case with 'non-negative', but I may be wrong.

Author: I interpret this Dragon behaviour as causing an accessibility failure in some cases that this rule passed.  I don't think it's practical for this rule to contain Dragon's dictionary and keep it up to date.  But I think it indicates an opportunity for a vendor to do it, and thereby excel.

    - For a) Judgment of "non-text" probably can't be fully automated.  For example: "X" for "close" probably can be automated, but presumably there are more cases than this.
    - For b) Use the Unicode classes Letter, Mark, and "Number, Decimal Digit [Nd]". (This will exclude hyphens, punctuation, emoji, and more.)

> - Remove all characters that are within parentheses (AKA round brackets).

Reviewer 2: Parenthesis mean something very different in German, so maybe shouldn't be removed in German text:
https://translationpost.com/2019/08/06/difference-german-english-parenthesis-usage/

And in Arabic parenthesis are often uses as quote marks:
https://forum.wordreference.com/threads/parentheses-in-arabic.1772289/

In accounting, numbers in round brackets are negative - typically losses or debts - so these are really important.
e.g. Profit: (2,400) is a 2,400 loss, but would match aria-label='Profit'

Author: That was educational.  I admit that this PR will not handle those cases well.  But at least it will handle them as "false negatives", which are inevitable in ACT rules.  I think that this concern - and all of your concerns here - are valuable, but I don't think that this PR can reasonably be expected to handle them all.  This PR "closes" 5 issues, explicitly "does not handle" 3 more, and (I think) doesn't do any harm.  I'd like to improve the rule a little, then take it from there.  This PR has been open for 10 months.  "Perfect is the enemy of good" and all that.

Reviewer 4: Given that we know this is an issue, I think it should be included some how as a note that it may result in false negatives in these situations. Particularly since we are already recognising i18n differences in the whitespace splitting part of this algorithm.

    - Ignore square brackets and braces.

> - Split the string into a list of strings, using a greedy [whitespace][] regular expression as the separator.

Reviewer 2: This part seems too implementation specific - normalising the strings by trimming and collapsing runs of [whitespace][] then looking for a substring match does the same thing (and is probably faster). Is there a way to write the algorithm in a more general way? (Counter example: the algorithms in specs like the URL Standard are very detailed and describe incrementing pointers)

> - Split the string into a list of strings, using a greedy [whitespace][] regular expression as the separator.

Reviewer 2: This part depends on your implementation language having a split on regex method. It's also quite slow - in Rust [regex::split](https://docs.rs/regex/latest/regex/struct.Split.html) has time-complexity O(m * n^2). Regexs in other languages are also very slow see: https://swtch.com/~rsc/regexp/regexp1.html

Normalizing the strings by trimming and collapsing runs of [whitespace][] then looking for a substring match does the same thing, and is much faster - you can normalize a string in O(n) time, and the time-complexity of substring matching algorithms is usually O(n-m+1). 

The algorithms in specs like the URL and HTML Standards are very detailed, but they only use very fast universally available operations like incrementing an index or comparing characters (which compile or JIT to single CPU instructions). If the algorithm was written in this style it would look something like:

```
A label is contained in name if:
1. Let normalizedLabel be the result of running voice-control string normalizer on label
2. Let normalizedName be the result of running voice-control string normalizer on name
3. If normalizedLabel matches a substring in normalizedName return true
4. Otherwise return false

The voice-control string normalizer takes a string input, and returns a normalized string:
1. Let output be the empty string
2. Let prev be the null code point
3. For each code point c in input:
 3.1 If c is non-text or c is whitespace let c = U+0020 Space // replace non-text with space characters
 3.2 If c == U+0020 Space
  3.2.1 If output is empty continue; // discard leading whitespace
  3.2.2 Otherwise If prev == U+0020 Space continue; // merge runs of space
 3.3 Let prev = c
 3.4 Append c to output
5. If prev == U+0020 Space remove the last code point from output // trim trailing whitespace
6. Return output
```

The algorithm should take Unicode normalization into account somewhere - the code point U+00F1 ñ [LATIN SMALL LETTER N WITH TILDE]  produces an identical glyph on screen to the code points U+006E n [LATIN SMALL LETTER N] followed by U+0303 ◌̃ [COMBINING TILDE]. There are specific algorithms to produce strings suitable for comparing: https://en.wikipedia.org/wiki/Unicode_equivalence. These are typically available in a library somewhere (e.g. String.prototype.normalize() in JS)

There are also tricky corner cases, like `ß` and `ss` being equivalent and voiced exactly the same in German (this might be handled by  unicode normalization)

Author: Good call out.  I'm going to keep defending regular expressions for the time being, because I can't see how to handle this case otherwise: `<a href="#" aria-label="Discover Italy">Discover It</a>`.   [Edit: I can see now, by adapting your string normalization pseudocode to output a list of strings instead of a string.  But the necessity of doing that is still unclear to me.]

Obviously we want to avoid a situation like "[over sixty seconds to match a 29-character string](https://swtch.com/~rsc/regexp/regexp1.html)".  It seems that the article's concern only applies to regular expressions which involve backtracking, and my algorithm's regex ("a greedy whitespace regular expression") isn't one of those.  In Javascript, my algorithm's regex would be this: `const regex = /\p{White_Space}+/gu;`, which (I think) can't result in any backtracking, because there is nothing before the "atom" `\p{White_Space}` to backtrack to.  

As for the concern about O(m * n^2) complexity, I'm unclear on what exactly m and n are, but I'm open to doing some tests.   As always with Big O notation, even if the scaling (for large m and n) is bad, the unspoken constant "C" might be small enough that it's not a real problem for our uses cases.  

Regarding Unicode normalization: no argument there.  I'll try to add something to that end.  Not sure when the best time to do that would be.

Reviewer 2: Thinking about this a bit more, a more serious issue is the whitespace RegEx and my suggested algorithm only work for a restricted set of languages. Thai, Chinese and Japanese don't use whitespace to separate words.

MDN also recommends against using Regex for word breaking:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Regular_expressions/Word_boundary_assertion#examples

One way of resolving this - instead of: 

> Split the string into a list of strings, using a greedy [whitespace][] regular expression as the separator.

say something like

> Split the string into a list of strings, using the word breaking rules for the inherited programmatic language.

This won't force a change on implementations already using regexes for European languages, but avoids incompatibility with languages like Thai, Chinese and Japanese. 

Of note: CSS doesn't specify exactly where word boundaries occur in different languages but does link to additional specs that do: https://drafts.csswg.org/css-text/#soft-wrap-opportunity

Author: Sorry I haven't had time to work on this.   I think I will later this month.

---
title: Label in Name Algorithm
key: label-in-name-algorithm
unambiguous: true
objective: false
input_aspects:
  - CSS styling
  - DOM tree
---

To check whether an [element][] has <dfn>its label contained in its name</dfn>, follow this algorithm:

Let 'label' be the [visible inner text][] of the target element.  Let 'name' be the [accessible name][] of the target element.  Both 'label' and 'name' are strings.

> Sub-algorithm to tokenize a string:

Reviewer 5: Since these steps are sequential, suggest they be numbered instead of bullet points.

Author: I just did this in this commit: https://github.com/act-rules/act-rules.github.io/commit/cd69eaf44932c3988893e88d70fed2e325efc84c

    - Ignore square brackets and braces.
- Split the string into a list of strings, one string per word, according to the word segmentation rules for the inherited programmatic language.
    - This 'split' operation must:
        - Effectively remove leading and trailing [whitespace][].
        - If the input string contains nothing but [whitespace][] before this operation: return an empty list.
    - In English and most other European languages, a greedy [whitespace][] regular expression will accomplish this.  In languages such as Thai, Chinese, and Japanese, it won't.
    - A consequence of using the ACT definition of [whitespace][] here is that all kinds of whitespace are covered.  That includes the Unicode code point U+00A0 - the "No-Break Space" - which can be represented by the HTML named character reference `&nbsp;`.

Then do the check: is the tokenized 'label' a sublist of the tokenized 'name'?
- This 'sublist' check has these properties:
    - It checks whether elements are consecutive or not.  That is: it checks for a substring, in the computer science sense of the term.  Not a subsequence.
    - An empty list is a sublist of any list.

> If the answer is "yes" (that is: the tokenized 'label' is a sublist of the tokenized 'name'), then this algorithm returns "is contained".  Otherwise, it returns "is not contained".

Reviewer 4: Daft question time. What if the label is 'This is a cat' and the name is 'This is not a cat'. The tokenized name would include the tokenized label but is obviously entirely incorrect.

I would be happy to be missing something in the specific definitions of 'is contained' and 'sublist'.

Author: @iadawn when you wrote "these situations", which situation do you mean?  Is it (a) (Judgment of "non-text")?  If so: I think that would result in a false positive (i.e. incorrect failing of the rule), not a false negative.

As for the 'This is a cat' case: I tried to cover that with the bullet point that starts with "It checks whether elements are consecutive or not."

Reviewer 4: 'These situations' relates to the situation where removal of parenthesis results in a change in meaning for certain languages. You [note that these will result in a false negative](https://github.com/act-rules/act-rules.github.io/pull/2075/files/7b2a05308cd402addfc88253a3b3d65ed7448c9d#r1566493727) but it might be worth being clear about that possibility in the text somehow.

Regarding the cat and rereading it I think that the problem may start with the statement:

> is the tokenized 'label' a sublist of the tokenized 'name'

That introduces the term 'sublist' - which has it's own particular meaning - and then defines that term in a different way to it's own particular meaning.

Perhaps another way to write this would be:

```suggestion
Then check if the consecutive tokenized 'label' is contained within the tokenized 'name'.
```

Or some such.

Author: > 'These situations' relates to the situation where removal of parenthesis results in a change in meaning for certain languages. 

@iadawn I see now.  Thank you for clarifying.  I just pushed two commits for it.  Does that cover it?

Author: > That introduces the term 'sublist' - which has it's own particular meaning - and then defines that term in a different way to it's own particular meaning.

You replaced "sublist" with "contained within".  In all respect, I think that's a step backwards.  "Contained within" is ambiguous, as was revealed by https://github.com/act-rules/act-rules.github.io/issues/1458 , which this PR (in it's current state i.e. using my definition of "sublist") fixes.  

> That introduces the term 'sublist' - which has it's own particular meaning - and then defines that term in a different way to it's own particular meaning.

I'm afraid I don't understand either part of that of that sentence.  1) Assuming that by "particular meaning" you mean "official definition": I couldn't find any official definition of "sublist", so I made my own.  2) Even if "sublist" had an official definition somewhere, I assume that official definition would define it to work the same as every programming language's "subList()" function, which means: consecutive elements, same order.  Which is the same way that my definition works.  So how my definition differs from that is beyond me.

---
title: Label in Name Algorithm
key: label-in-name-algorithm
unambiguous: true
objective: false
input_aspects:
  - CSS styling
  - DOM tree
---

To check whether an [element][] has <dfn>its label contained in its name</dfn>, follow this algorithm:

> Let 'label' be the [visible inner text][] of the target element.  Let 'name' be the [accessible name][] of the target element.  Both 'label' and 'name' are strings.

Reviewer: ```suggestion
Let `label` be the [visible inner text][] of the target element.  Let `name` be the [accessible name][] of the target element.  Both `label` and `name` are strings.
```

Sub-algorithm to tokenize a string:

1. Do Unicode [case folding][] on the string then convert it to [normalization form KD][].

> 1. For each character that either a) represents non-text content, or b) isn't a letter or a digit: replace that character with a space character.

Reviewer 6: I understand that there is (or should be) a separate issue addressing non-latin languages, but IMO the current wording is overly strict. Japanese for example is not composed of letters. As a result, a string like "これは何ですか", per this bullet, gets replaced with seven whitespace characters, which is clearly incorrect and leads to invalid results. Given that the current/existing rule handles simple Japanese scenarios but the new one does not, IMO we can't proceed in this direction. We need a better definition of "letter or digit" that also accounts for ideographic writing systems; otherwise, this change would be a step backward.

Author: Fair point for sure, but I'm a little stumped as to how to fix this.  Do you have any ideas?

Author: Wait a minute.  I think that Japanese works out fine because they are "letters" according to Unicode.  For example, see https://www.compart.com/en/unicode/U+3059 - it says "Category: Other Letter (Lo)".  That is covered by my algorithm and won't be replaced by a space.  I just pushed [a commit](https://github.com/dan-tripp-siteimprove/act-rules.github.io/commit/60e573459951c3af7e8a8034009130a189a7539f) which changes some small things, but does not change this case.

>     - For a) Judgment of "non-text" probably can't be fully automated.  For example: "X" for "close" probably can be automated, but presumably there are more cases than this.

Reviewer: **Issue:** I'm not a huge fan of saying something can or cannot be automatised. How about turning it in an open list of cases: "characters that are non-text include, but are not restricted to, emoji, the use of 'X' meaning 'close', punctuation (P Unicode class), …"

Author: I would like to discuss this one a little more, along the lines of one of my earlier comments about hard-to-read specs and how I don't want to add to that situation.

If an algorithm becomes hand-wavey in a section - like mine does, right here - then IMO it should say so.

If your lack of fandom is due to the fact that "probably can't be fully automated" is probably not true because one could likely build a local ML model today that out-performs most humans at this job ... then I agree.  Then we need a succinct way of saying _that_ ... without glossing over the fact that we just entered a hand-wavey section.

Giacomo recently made some comments on manual vs. automated, referenced here - https://github.com/act-rules/act-rules.github.io/issues/2370 .

Reviewer: Yes, I agree with your analysis. That is why I was proposing the wording "include but is not restricted to" which both signals a hand-wavy part, gives room for several examples, and can easily be extended at a later point. It can of course be made like "(test) requires some judgment call, it contains but is not restricted to A, B, C, …"

Author: I made an attempt at this in 1e151d59c0bdb2d125f4444c82e98bd8c8377253.

>     - For b) Use the Unicode general categories "L" (Letter) and "N" (Number).  (This will exclude hyphens, punctuation, emoji, and more.)

Reviewer: **Suggestion:** Can you add links to the definition of the Unicode categorise?

Author: I did it in commit bf9bf6a13202877ece446cb52b7e7fea0353ed88.

> 1. Remove parentheses (also known as round brackets) and all characters that are between an opening and closing parenthesis.

Reviewer: **Question:** Is that implicitly assuming that parenthesis are well-balanced? How does the algorithm handle "Enter date (YYYY-MM-DD" or "Enter Date (YYYY-MM-DD))" or "Enter date (YYYY-MM-DD) of birth)"

Author: Yes :)  but it's not so bad, because this rule considers inapplicable any example like that - eg. `<label>Enter date (YYYY-MM-DD<input aria-label="Date" type="text"></label>` - because the rule's applicability covers only [elements supporting name from content](https://www.w3.org/TR/wai-aria-1.2/#namefromcontent), and `<input>` doesn't.

Reviewer: Ah, yes 😅 

But we can make the same thing with links, maybe just a bit more contrived, but something like a Wiki disambiguation: `Equator (country)`, …

Author: That's not contrived.  I made an attempt in commit 64fe7793209fd1b03b870b196062ee8c19d7528a.

Author: ... not an attempt to handle unbalanced parenthesis.   An attempt to handle "stuff in parenthesis in English which should not be ignored".  I still don't know how to handle unbalanced parenthesis.

> 1. Remove parentheses (also known as round brackets) and all characters that are between an opening and closing parenthesis.

Reviewer: **Issue:** [U+0028 LEFT PARENTHESIS](https://unicode-explorer.com/c/0028) is not in categories L or N. So it has already been removed by Step 1b. (as well as other kind of brackets that are in the Ps/Pe categories)

Author: Yikes.  I just fixed it in commit e6dc5c88a5d9a40e0b66bf7cc35f06568601a54a, I think.  This problem has been there since the start of this PR (June 2023) and you're the first to notice.

> 1. Remove parentheses (also known as round brackets) and all characters that are between an opening and closing parenthesis.

Reviewer 6: We need to exclude anything that is a mathematical, regular, logical, or similar type of expression.
For example, in a multiple-answer quiz asking "Which of the following is correct?":
- if ((a || b) && c)
- if (a || b && c)
- …

Here the parentheses make a difference so treating them as having the same accessible name would be invalid

Author: It's another fine puzzle.   How could we exclude such things?  Change this:
`Remove parentheses (also known as round brackets) and all characters that are between an opening and closing parenthesis.`
to this? 
`Remove parentheses (also known as round brackets) and all characters that are between an opening and closing parenthesis, unless the parentheses are part of a mathematical, regular, logical, or similar type of expression.`

>     - Don't do this for square brackets, nor braces.

Reviewer: ```suggestion
    - Don't do this for other kind of brackets such as square brackets, braces, …
```

> 1. Split the string into a list of strings, one string per word, according to the word segmentation rules for the inherited programmatic language.

Reviewer: **Suggestion:** We should have a link for "inherited programmatic language" (of the element), which I realise is tricky because [the existing definition](https://github.com/act-rules/act-rules.github.io/blob/develop/pages/glossary/text-inheriting-language.md) goes the other way around.

Author: Oh dear.  Would switching to [most common element language](https://www.w3.org/WAI/standards-guidelines/act/rules/terms/element-language/examples/#most-common-element-language) help?

Reviewer: 🤔 I guess so, it depends whether we want to act on the `lang` attribute or the actual language…

OTOH, it looks like [HTML definition of language](https://html.spec.whatwg.org/multipage/dom.html#language) is already what we need 😄

Author: Right on - I just did it in commit 3586285941915371fdb9781b8685a01369483ae5

    - This 'split' operation must:
        - Effectively remove leading and trailing [whitespace][].
        - If the input string contains nothing but [whitespace][] before this operation: return an empty list.

>     - In English and most other European languages, a greedy [whitespace][] regular expression will accomplish this.  In languages such as Thai, Chinese, and Japanese, it won't.

Reviewer: ```suggestion
    - In English and most other European languages, a greedy [whitespace][] regular expression will accomplish this. In languages such as Thai, Chinese, and Japanese, it won't.
```

    - A consequence of using the ACT definition of [whitespace][] here is that all kinds of whitespace are covered.  That includes the Unicode code point U+00A0 - the "No-Break Space" - which can be represented by the HTML named character reference `&nbsp;`.

> Then do the check: is the tokenized 'label' a sublist of the tokenized 'name'?

Reviewer: ```suggestion
Then do the check: is the tokenized `label` a sublist of the tokenized `name`?
```

>     - A consequence of using the ACT definition of [whitespace][] here is that all kinds of whitespace are covered.  That includes the Unicode code point U+00A0 NO-BREAK SPACE (NBSP), which can be represented by the HTML named character reference `&nbsp;`.

Reviewer: ```suggestion
    - A consequence of using the ACT definition of [whitespace][] here is that all kinds of whitespace are covered. That includes the Unicode code point U+00A0 NO-BREAK SPACE (NBSP), which can be represented by the HTML named character reference `&nbsp;`.
```

- This 'sublist' check has these properties:

>     - Each string comparison (between a list element in the tokenized label and a list element in the tokenized name) is a simple string equality check.

Reviewer: **Note:** "Contiguous subsequence" or "subarray" seem to be the commonly used terms, but I do not find a reference that would be authoritative enough to be pointed here :-/ 

Maybe this comparison can be rephrased as "check whether the tokenised `label` list can be obtained by removing any number of tokens from the start and/or end of the tokenised `name` list"?

Author: I could go for that, if I can still say "sublist" [1], and then define "sublist" as: "list X can be obtained by removing...")    The alternative is to make the sentence "Then do the check: is the tokenized 'label' a sublist of the tokenized 'name'" long by stuffing the "definition of sublist" into it.  I don't think that's fair to any reader who doesn't have a Computer Science degree, and that includes me. 

[1] Replace "sublist" with anything, as long as it's one word.

Reviewer: We can also move this definition to its own definition in our glossary, so that this one entry is shorter. But maybe that makes too many redirections… (I agree we do need the definition somewhere)

I think I prefer the term "contiguous subsequence" to "sublist" or "subarray". Because it does contain the "contiguous" part of it which is missing from other similar terms. And "subsequence" has some relatively well-spread meaning, e.g. in the classical "longest common subsequence" problem (a poster child of Dynamic Programming).

Author: I like "contiguous subsequence".  I just gave that a whirl in commit 020300cc7b60a9faaf74a4775deb3e36a91e6d4a.

Author: Oops: commit b01f57aa49af814b51942cd868ebad7516061d24.

Author: Oops again: commit 94fcd8c887266ad21b57db6fb2c7dd870a5ddc98.

>     - The "contiguous" aspect means that it's crucial that the elements are consecutive in the original list.  Put another way: a contiguous subsequence of X can be obtained by removing any number of tokens from the start and/or end (but not the middle) of X.  For example: ["A", "B", "C"] is a contiguous subsequence of ["A", "B", "C", "D"].  ["A", "B", "D"] is not.

Reviewer: ```suggestion
    - The "contiguous" aspect means that it's crucial that the elements are consecutive in the original list. Put another way: a contiguous subsequence of X can be obtained by removing any number of tokens from the start and/or end (but not the middle) of X. For example: ["A", "B", "C"] is a contiguous subsequence of ["A", "B", "C", "D"]; but ["A", "B", "D"] is not.
```

    - An empty list is a sublist of any list.

If the answer is "yes" (that is: the tokenized 'label' is a sublist of the tokenized 'name'), then this algorithm returns "is contained".  Otherwise, it returns "is not contained".

[accessible name]: #accessible-name 'Definition of accessible name'
[case folding]: https://www.w3.org/TR/charmod-norm/#dfn-case-folding
[element]: https://dom.spec.whatwg.org/#element
[normalization form KD]: https://www.unicode.org/glossary/#normalization_form_kd
[visible inner text]: #visible-inner-text 'Definition of Visible inner text'

> [whitespace][]: #whitespace 'Definition of whitespace'

Reviewer: ```suggestion
[whitespace]: #whitespace 'Definition of whitespace'
```
